documentation/adr/0018-algorithm-migration-to-paged-files.md

```diff
@@ -0,0 +1,115 @@
+
+There are several algorithms that need to be updated to support large datasets.
+Any update must be made in such a way that algorithms that the whole of Horace continues to function as it currently does on "small" datasets
+until a point at which all algorithms are updated.
+During the migration it is also necessary that scripts using a subset of the updated algorithms
+can be created to perform end-to-end testing.
+These should not require code changes or alternate APIs.
+
+
+## Decision
+
+- A new API for SQW object will be created.
+The SQW class will to default to in memory operation (i.e. no paging)
+if a page size is not passed explicitly to the constructor
+until all code has been updated to support paged operation
+
+```matlab
+% unpaged: default page size is 0/inf
+s = sqw(filename)
+
+% paged with specified page size
+s = sqw(filename, page_size)
+```
+- a warning message will be output when operations are performed
+on dataset that require paging to enable a user to understand the performance drop
+- all algorithms will be created with a common API pattern.
+The two APIs will share a common implementation,
+the filename variant creating a (paged) SQW object
+	- `algorithm(filename, outfilename)`
+perform operation on a new SQW object that  will be explicitly created as paged
+	- `algorithm(sqw, [outfilename])`
+perform operation on existing SQW object (may be paged or unpaged)
+
+The example APIs for `cut` would be:
+```matlab
+% Pure in-memory operation (small file only)
+s = sqw(infile) 
+slice = cut(s)
+
+% Paged operation (in-memory for small file, paged for large file)
+s = sqw(infile, page_size)
+slice = cut(x) 		 % object => object
+
+% Filename based API
+slice = cut(infile)  % disk => object
+cut(infile, outfile) % disk => disk
+```
+
+
+## Consequences
+
+- The internal switch between in-memory and disk operations means
+the performance drop will happen at the point the dataset exceeds page size.
+This switch will present as a sudden, unexpected performance drop.
+Display of a console message will aid users in understanding the change.
+- Paging will not be generally enabled.
+All SQW object will continue to work with existing (unmodified) algorithms
+- Paging can be switched “on” simply supporting developer activity
+- If paging is enabled, i.e. the filename API used, and the SQW file is less that page sized,
+the operation will be performed in memory giving improved performance
+- A single implementation of each algorithm will be create that will work for paged data,
+this will then work for in-memory or file backed use-cases
+- The paged SQW file created by the filename API must not be returned from the algorithm call
+to prevent "page leakage" and unexpected failures elsewhere in the application
+- Existing algorithms will continue to work on **unpaged data**.
+The behaviour if executed on **paged** data is undefined (should this throw an error?)
```

Comment:
Not sure what you mean by this. An algorithm that has not been updated for paged data should never be using paged data. Paged `PixelData` objects are not publicly exposed, so we shouldn't run into any problems like this.

Reply:
I added that as a consequence of the other strategy.

If it does happen, for example a developer calling with a paged object, or an bug in a paged algorithm implementation that leaks the paged file, we need to raise that be behaviour is undefined.

Raising an error would be a nicer solution, but would require code changes everywhere.